Mathlib/Analysis/SpecialFunctions/AGM.lean

```diff
@@ -0,0 +1,242 @@
+  exact i₂.trans i₃
+
+lemma le_agmStep_iterate_snd (n : ℕ) : g ≤ (agmStep^[n] (g, a)).2 := by
+  obtain ⟨i₁, i₂, -⟩ := agmStep_iterate_bounds h n
+  exact i₁.trans i₂
+
+lemma bddAbove_range_agmStep_iterate_fst : BddAbove (Set.range fun n ↦ (agmStep^[n] (g, a)).1) := by
+  rw [bddAbove_def]
+  exact ⟨a, by simpa using agmStep_iterate_fst_le h⟩
+
+lemma bddBelow_range_agmStep_iterate_snd : BddBelow (Set.range fun n ↦ (agmStep^[n] (g, a)).2) := by
+  rw [bddBelow_def]
+  exact ⟨g, by simpa using le_agmStep_iterate_snd h⟩
+
+end
+
+lemma agmStep_iterate_zero {n : ℕ} : agmStep^[n] (0, a) = (0, a * 2⁻¹ ^ n) := by
+  induction n with
+  | zero => simp
+  | succ n ih =>
+    simp_rw [iterate_succ', comp_apply, agmStep, ih, zero_mul, sqrt_zero, zero_add]
+    congr
+    ring
+
+lemma agmStep_iterate_mul {k : ℝ≥0} {n : ℕ} :
+    agmStep^[n] (k * g, k * a) = k • agmStep^[n] (g, a) := by
+  induction n with
+  | zero => simp
+  | succ n ih =>
+    simp_rw [iterate_succ', comp_apply, ih, agmStep, Prod.smul_fst, Prod.smul_snd]
+    rw [smul_mul_smul, smul_eq_mul, sqrt_mul, sqrt_mul_self, ← smul_eq_mul,
+      ← smul_add, smul_div_assoc, Prod.smul_mk]
+
+open Filter Topology
+
+lemma tendsto_agmStep_iterate_zero :
+    Tendsto (fun n ↦ (agmStep^[n] (0, a)).1) atTop (𝓝 0) ∧
+    Tendsto (fun n ↦ (agmStep^[n] (0, a)).2) atTop (𝓝 0) := by
+  simp_rw [agmStep_iterate_zero, tendsto_const_nhds_iff]
+  rw [true_and, ← mul_zero a]
+  exact (tendsto_pow_atTop_nhds_zero_of_lt_one (by norm_num)).const_mul a
+
+/-- Iterating `agmStep` drives both components towards a common value. -/
+lemma exists_tendsto_agmStep_iterate :
+    ∃ M, Tendsto (fun n ↦ (agmStep^[n] (g, a)).1) atTop (𝓝 M) ∧
+      Tendsto (fun n ↦ (agmStep^[n] (g, a)).2) atTop (𝓝 M) := by
+  wlog h : g ≤ a
+  · rw [not_le] at h
+    obtain ⟨M, hM₁, hM₂⟩ := this h.le
+    refine ⟨M, hM₁.congr' ?_, hM₂.congr' ?_⟩
+    all_goals
+      rw [EventuallyEq, eventually_atTop]
+      refine ⟨1, fun n hn ↦ ?_⟩
+      rw [agmStep_iterate_comm (Nat.ne_zero_of_lt hn)]
+  rcases (zero_le g).eq_or_lt with gzero | gpos
+  · subst gzero
+    exact ⟨0, tendsto_agmStep_iterate_zero⟩
+  have bdd_g := bddAbove_range_agmStep_iterate_fst h
+  have tendsto_g := tendsto_atTop_ciSup (agmStep_iterate_fst_monotone h) bdd_g
+  set M := ⨆ n, (agmStep^[n] (g, a)).1
+  refine ⟨M, tendsto_g, ?_⟩
+  have a_rearrangement (n : ℕ) :
+      (agmStep^[n] (g, a)).2 = (agmStep^[n + 1] (g, a)).1 ^ 2 / (agmStep^[n] (g, a)).1 := by
+    simp_rw [iterate_succ', comp_apply, agmStep, sq_sqrt]
+    rw [mul_div_cancel_left₀ _ (gpos.trans_le (agmStep_iterate_bounds h n).1).ne']
+  simp_rw [a_rearrangement]
+  rw [← mul_self_div_self M, ← sq]
+  refine (Tendsto.pow ?_ 2).div tendsto_g (gpos.trans_le (le_ciSup bdd_g 0)).ne'
+  exact (tendsto_add_atTop_iff_nat 1).mpr tendsto_g
+
+/-- The arithmetic-geometric mean of two `NNReal`s. -/
+noncomputable def agm (x y : ℝ≥0) : ℝ≥0 :=
+  (@exists_tendsto_agmStep_iterate x y).choose
```

Comment:
Here I would connect the sequence information to `agm` using `choose_spec`. I think it's useful to have those lemmas.